### DISCUSSION TITLE
Literals in SELECT columns incorrectly converted to strings

### DISCUSSION THREAD

**Discussion by TallJimbo** [sqlalchemy/sqlalchemy#8540] [Usage Questions] [answered]
### Describe the bug

PostgreSQL range literals that appear in the SELECT clause (i.e. result columns) of a query are converted to strings as they would be when actually being passed to the database, but are not converted back to `psycopg2.extras.NumericRange` instances as result values usually are.

### To Reproduce

```python
import sqlalchemy.dialects.postgresql
from psycopg2.extras import NumericRange


engine = sqlalchemy.engine.create_engine("postgresql://user:pass@localhost:5432/db")

query = sqlalchemy.sql.select(
    [
        sqlalchemy.sql.literal(
            NumericRange(lower=4, upper=8), type_=sqlalchemy.dialects.postgresql.INT8RANGE
        ).label("a")
    ]
)
with engine.begin() as connection:
    row = connection.execute(query).fetchone()
assert isinstance(row.a, NumericRange), f"{type(row.a)} is not an instance of NumericRange"
```

### Error

Expected assertion not to be triggered, by either passing the literal through as-is or having it converted it back from a string to a `NumericRange`.

### Versions

- OS: Linux Mint 21
- Python: 3.10
- SQLAlchemy: 1.4.41
- Database: PostgreSQL
- DBAPI: psycopg2

### Additional context

I also see the problem when the literal is just one of several columns in the SELECT clause and the others are "real" columns from a real table.

From inspecting the query that is actually executed it seems the literal column is cleverly elided rather than actually included, but I haven't been able to track down where the literal value is instead injected into the result row objects, or why it's being converted one way rather than not at all or both ways (which would at least be correct, if inefficient).

Another possible factor is that this is really a psycopg2 custom type in terms of where the conversion to/from strings happens (the SQLAlchemy type wrapper only exists to define some operators).

**Comment by zzzeek** [comment, marked as answer]
the PG drivers do the range there and they pass them as strings.  The database has no way to know , returning it back, that this is not a string.  so you need to cast things like this:

```py
query = sqlalchemy.sql.select(
        sqlalchemy.cast(sqlalchemy.sql.literal(
            NumericRange(lower=4, upper=8),
        ), type_=sqlalchemy.dialects.postgresql.INT8RANGE).label("a")
)
```

same goes for other "composed" values like JSON, HSTORE, etc.   the database doesn't know what these are without a database table column present.

**Reply by TallJimbo** [reply]
Thanks; I wasn't aware of `cast`, and thought I had provided that type information by passing `type_` to `literal`.  I'm guessing that would have worked for a type whose conversions were managed directly by SQLAlchemy `TypeEngine` hooks, while the `cast` is necessary in this case to invoke `psycopg2`'s?  Is there any utility to also passing `type_` to literal in the solution you provided?

**Reply by zzzeek** [reply]
SQLAlchemy renders driver-specific CAST automatically in some cases but for this case I tested a few types where this kind of thing is sensitive (like JSON) and for the straight up literal() with no other context, it doesn't render any CAST, nor for JSON either.   

In 2.0 there is a simple boolean for this so I can likely switch it on for the PG RANGE types.   Note that in 2.0, your code may have to change here as RANGEs are all new:

https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#range-and-multirange-types

**Reply by zzzeek** [reply]
I have a review for autocast for 2.0 here: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4089

we dont seem to do this in general for other types, which maybe we should.   at least this serves as a POC if we wanted to turn this on for other types that are usually passed to the DB as strings like JSON; it's also not clear if that would be a PG only thing or if it applies to other backends too.  So ranges are at least PG specific.
